doc/util/chplcheck-docs.py

```diff
@@ -0,0 +1,205 @@
+    def rst(self):
+        lines = []
+        lines.append(self.name)
+        lines.append("~" * len(self.name))
+        lines.append("")
+        lines.append(
+            "Is enabled by default? " + ("Yes" if self.default else "No")
+        )
+        lines.append("")
+
+        if self.description:
+            lines.append(self.description)
+            lines.append("")
+
+        if self.settings:
+            lines.append("Settings:")
+            for setting in self.settings:
+                lines.append(f"  - ``{setting}``")
+            lines.append("")
+
+        if self.example_text:
+            lines.append(self.example_text)
+            lines.append("")
+
+        return "\n".join(lines)
+
+    def add_example(self, example_directory: str):
+        # find the example file
+        example_file = os.path.join(example_directory, self.name + ".chpl")
+        if not os.path.exists(example_file):
+            return
+
+        with open(example_file) as handle:
+            pieces = chpl2rst.to_pieces(handle, False)
+            rstoutput = chpl2rst.gen_rst(pieces, example_file)
+
+        self.example_text = rstoutput
+
+
+def find_rules(file: str):
+
+    def get_rule(func) -> typing.Optional[Rule]:
+        """
+        Given a function definition, return a Rule object if it is a rule, otherwise None
+        """
+        if not isinstance(func, ast.FunctionDef):
+            return None
+
+        # find all decorators by walking decorator_list and filtering for Attribute
+        decorators = [
+            i
+            for d in func.decorator_list
+            for i in ast.walk(d)
+            if isinstance(i, ast.Attribute)
+        ]
+
+        # if there is at least 1 decorator thats a `basic_rule`, an `advanced_rule`,
+        # or a `location_rule`, keep it
+        if not any(
+            d.attr in ["basic_rule", "advanced_rule", "location_rule"]
+            for d in decorators
+        ):
+            return None
+
+        # get the name of the rule
+        name = func.name
+        # get the docstring of the rule
+        description = ast.get_docstring(func) or ""
+
+        # get the patterns, if they exist
+        # the patterns are the first argument of the decorator
+        # we get the string representation of the pattern
+        patterns = [
+            ast.unparse(d.args[0])
+            for d in func.decorator_list
+            if isinstance(d, ast.Call)
+            and isinstance(d.func, ast.Attribute)
+            and len(d.args) > 0
+        ]
+
+        # determine if the rule is enabled by default
+        # default = any(d.attr == 'default' for d in decorators)
+        default_settings = [
+            ast.unparse(next(k.value for k in d.keywords))
+            for d in func.decorator_list
+            if isinstance(d, ast.Call)
+            and isinstance(d.func, ast.Attribute)
+            and len(d.keywords) > 0
+            and any(k.arg == "default" for k in d.keywords)
+        ]
```

Comment:
Can you build a dict of kwargs and check it instead of doing this iteration over and over?